File: langs/zig/src/runtime.zig
//! Clide runtime argument parser

const std = @import("std");
const spec = @import("spec.zig");

pub const ArgError = error{
    UnknownOption,
    MissingValue,
    MissingPositional,
    UnexpectedArgument,
    InvalidInt,
    InvalidBool,
    OutOfMemory,
};

pub const ParseResult = struct {
    command: []const u8,
    options: std.ArrayList(struct { []const u8, std.ArrayList([]const u8) }),
    positionals: std.ArrayList(struct { []const u8, []const u8 }),
    leftovers: std.ArrayList([]const u8),
    allocator: std.mem.Allocator,

    pub fn deinit(self: *ParseResult) void {
        for (self.options.items) |*opt| {
            opt.@"1".deinit();
        }
        self.options.deinit();
        self.positionals.deinit();
        self.leftovers.deinit();
    }
};

pub fn chooseCommand(spec_: *const spec.Spec, argv: []const []const u8) *const spec.Command {
    if (argv.len == 0) {
        return &spec_.commands[0];
    }

    const first_arg = argv[0];
    for (spec_.commands) |*cmd| {
        if (firstLiteral(cmd)) |lit| {
            if (std.mem.eql(u8, lit, first_arg)) {
                return cmd;
            }
        }
    }

    return &spec_.commands[0];
}

fn firstLiteral(cmd: *const spec.Command) ?[]const u8 {
    for (cmd.items) |item| {
        switch (item) {
            .required => |group| {
                switch (group) {
                    .single => |atom| {
                        switch (atom) {
                            .lit => |lit| return lit,
                            else => {},
                        }
                    },
                    .alt => |atoms| {
                        for (atoms) |atom| {
                            switch (atom) {
                                .lit => |lit| return lit,
                                else => {},
                            }
                        }
                    },
                }
            },
            .optional => {},
        }
    }
    return null;
}

pub fn parseWith(allocator: std.mem.Allocator, cmd: *const spec.Command, argv: []const []const u8) ArgError!ParseResult {
    var seq = std.ArrayList(struct { bool, spec.Group }).init(allocator);
    defer seq.deinit();

    for (cmd.items) |item| {
        const is_optional = switch (item) {
            .required => false,
            .optional => true,
        };
        const group = switch (item) {
            .required => |g| g,
            .optional => |g| g,
        };
        try seq.append(.{ is_optional, group });
    }

    var atoms = std.ArrayList(spec.Atom).init(allocator);
    defer atoms.deinit();

    for (seq.items) |item| {
        const group = item.@"1";
        switch (group) {
            .single => |atom| try atoms.append(atom),
            .alt => |alts| {
                for (alts) |atom| {
                    try atoms.append(atom);
                }
            },
        }
    }

    // Build list of known options
    var known_opts = std.ArrayList([]const u8).init(allocator);
    defer known_opts.deinit();

    for (atoms.items) |atom| {
        switch (atom) {
            .opt_bool => |opt| {
                if (opt.long) |long| try known_opts.append(long);
                if (opt.short) |short| try known_opts.append(short);
            },
            .opt_val => |opt| {
                if (opt.long) |long| try known_opts.append(long);
                if (opt.short) |short| try known_opts.append(short);
            },
            else => {},
        }
    }

    // Build list of positionals
    var pos_list = std.ArrayList(struct { []const u8, spec.Type, ?[]const u8 }).init(allocator);
    defer pos_list.deinit();

    for (atoms.items) |atom| {
        switch (atom) {
            .pos => |pos| {
                try pos_list.append(.{ pos.name, pos.ty, pos.default });
            },
            else => {},
        }
    }

    var opts_map = std.ArrayList(struct { []const u8, std.ArrayList([]const u8) }).init(allocator);
    var pos_map = std.ArrayList(struct { []const u8, []const u8 }).init(allocator);
    var seen_pos: usize = 0;
    var leftovers = std.ArrayList([]const u8).init(allocator);

    var i: usize = 0;
    while (i < argv.len) {
        const arg = argv[i];

        if (std.mem.eql(u8, arg, "--")) {
            // Handle remaining positionals
            const remaining_pos = pos_list.items[seen_pos..];
            const rest = argv[(i + 1)..];

            for (remaining_pos, 0..) |pos_item, j| {
                if (j < rest.len) {
                    const val = try parseVal(allocator, pos_item.@"1", rest[j]);
                    const name_owned = try allocator.dupe(u8, pos_item.@"0");
                    try pos_map.append(.{ name_owned, val });
                    seen_pos += 1;
                } else if (pos_item.@"2") |default| {
                    const val = try parseVal(allocator, pos_item.@"1", default);
                    const name_owned = try allocator.dupe(u8, pos_item.@"0");
                    try pos_map.append(.{ name_owned, val });
                    seen_pos += 1;
                } else {
                    return error.MissingPositional;
                }
            }

            for (rest[remaining_pos.len..]) |leftover| {
                const leftover_owned = try allocator.dupe(u8, leftover);
                try leftovers.append(leftover_owned);
            }
            break;
        } else if (arg.len > 0 and arg[0] == '-') {
            const eq_pos = std.mem.indexOfScalar(u8, arg, '=');
            const name_part = if (eq_pos) |pos| arg[0..pos] else arg;
            const val_part = if (eq_pos) |pos| arg[(pos + 1)..] else null;

            var found = false;
            for (known_opts.items) |opt| {
                if (std.mem.eql(u8, opt, name_part)) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                return error.UnknownOption;
            }

            // Find the option declaration
            var opt_decl: ?spec.Atom = null;
            for (atoms.items) |atom| {
                switch (atom) {
                    .opt_val => |opt| {
                        if ((opt.long != null and std.mem.eql(u8, opt.long.?, name_part)) or
                            (opt.short != null and std.mem.eql(u8, opt.short.?, name_part)))
                        {
                            opt_decl = atom;
                            break;
                        }
                    },
                    .opt_bool => |opt| {
                        if ((opt.long != null and std.mem.eql(u8, opt.long.?, name_part)) or
                            (opt.short != null and std.mem.eql(u8, opt.short.?, name_part)))
                        {
                            opt_decl = atom;
                            break;
                        }
                    },
                    else => {},
                }
            }

            if (opt_decl == null) {
                return error.UnknownOption;
            }

            switch (opt_decl.?) {
                .opt_bool => {
                    const name_owned = try allocator.dupe(u8, name_part);
                    const val_owned = try allocator.dupe(u8, "true");
                    addOpt(allocator, &opts_map, name_owned, val_owned) catch |e| {
                        allocator.free(name_owned);
                        allocator.free(val_owned);
                        return e;
                    };
                    i += 1;
                },
                .opt_val => |opt| {
                    const val = if (val_part) |v|
                        try parseVal(allocator, opt.ty, v)
                    else if (i + 1 < argv.len) {
                        i += 1;
                        try parseVal(allocator, opt.ty, argv[i])
                    } else {
                        return error.MissingValue;
                    };
                    const name_owned = try allocator.dupe(u8, name_part);
                    addOpt(allocator, &opts_map, name_owned, val) catch |e| {
                        allocator.free(name_owned);
                        return e;
                    };
                    i += 1;
                },
                else => unreachable,
            }
        } else {
            // Try to consume as literal
            var consumed = false;
            for (seq.items) |item| {
                if (item.@"0") { // optional
                    const group = item.@"1";
                    switch (group) {
                        .single => |atom| {
                            switch (atom) {
                                .lit => |lit| {
                                    if (std.mem.eql(u8, lit, arg)) {
                                        consumed = true;
                                        break;
                                    }
                                },
                                else => {},
                            }
                        },
                        .alt => |alts| {
                            for (alts) |atom| {
                                switch (atom) {
                                    .lit => |lit| {
                                        if (std.mem.eql(u8, lit, arg)) {
                                            consumed = true;
                                            break;
                                        }
                                    },
                                    else => {},
                                }
                            }
                            if (consumed) break;
                        },
                    }
                }
            }

            if (consumed) {
                i += 1;
                continue;
            }

            // Treat as positional
            if (seen_pos >= pos_list.items.len) {
                return error.UnexpectedArgument;
            }

            const pos_item = pos_list.items[seen_pos];
            const val = try parseVal(allocator, pos_item.@"1", arg);
            const name_owned = try allocator.dupe(u8, pos_item.@"0");
            try pos_map.append(.{ name_owned, val });
            seen_pos += 1;
            i += 1;
        }
    }

    // Add option defaults
    for (atoms.items) |atom| {
        switch (atom) {
            .opt_val => |opt| {
                const key = opt.long orelse opt.short orelse continue;
                var found = false;
                for (opts_map.items) |opt_item| {
                    if (std.mem.eql(u8, opt_item.@"0", key)) {
                        found = true;
                        break;
                    }
                }
                if (!found) {
                    if (opt.default) |default| {
                        const val = try parseVal(allocator, opt.ty, default);
                        const key_owned = try allocator.dupe(u8, key);
                        addOpt(allocator, &opts_map, key_owned, val) catch |e| {
                            allocator.free(key_owned);
                            return e;
                        };
                    }
                }
            },
            .opt_bool => |opt| {
                const key = opt.long orelse opt.short orelse continue;
                var found = false;
                for (opts_map.items) |opt_item| {
                    if (std.mem.eql(u8, opt_item.@"0", key)) {
                        found = true;
                        break;
                    }
                }
                if (!found) {
                    const val_owned = try allocator.dupe(u8, "false");
                    const key_owned = try allocator.dupe(u8, key);
                    addOpt(allocator, &opts_map, key_owned, val_owned) catch |e| {
                        allocator.free(key_owned);
                        allocator.free(val_owned);
                        return e;
                    };
                }
            },
            else => {},
        }
    }

    // Add positional defaults
    for (pos_list.items) |pos_item| {
        var found = false;
        for (pos_map.items) |pos_map_item| {
            if (std.mem.eql(u8, pos_map_item.@"0", pos_item.@"0")) {
                found = true;
                break;
            }
        }
        if (!found) {
            if (pos_item.@"2") |default| {
                const val = try parseVal(allocator, pos_item.@"1", default);
                const name_owned = try allocator.dupe(u8, pos_item.@"0");
                try pos_map.append(.{ name_owned, val });
            } else {
                return error.MissingPositional;
            }
        }
    }

    return ParseResult{
        .command = cmd.name,
        .options = opts_map,
        .positionals = pos_map,
        .leftovers = leftovers,
        .allocator = allocator,
    };
}

fn parseVal(allocator: std.mem.Allocator, ty: spec.Type, s: []const u8) ArgError![]const u8 {
    return switch (ty) {
        .int => {
            _ = std.fmt.parseInt(i64, s, 10) catch return error.InvalidInt;
            const owned = try allocator.dupe(u8, s);
            return owned;
        },
        .bool => {
            var lower_buf: [256]u8 = undefined;
            const lower = std.ascii.lowerString(&lower_buf, s);
            if (std.mem.eql(u8, lower, "true") or std.mem.eql(u8, lower, "false")) {
                const owned = try allocator.dupe(u8, lower);
                return owned;
            } else {
                return error.InvalidBool;
            }
        },
        .str, .path => {
            const owned = try allocator.dupe(u8, s);
            return owned;
        },
    };
}

fn addOpt(allocator: std.mem.Allocator, opts_map: *std.ArrayList(struct { []const u8, std.ArrayList([]const u8) }), key: []const u8, val: []const u8) ArgError!void {
    // key and val are already owned (duplicated) by the caller
    for (opts_map.items) |*opt| {
        if (std.mem.eql(u8, opt.@"0", key)) {
            try opt.@"1".append(val);
            return;
        }
    }
    var vals = std.ArrayList([]const u8).init(allocator);
    try vals.append(val);
    try opts_map.append(.{ key, vals });
}

File: langs/rust/src/runtime.rs
//! Clide runtime argument parser

use crate::spec::*;

#[derive(Debug)]
pub struct ArgError(pub String);

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParseResult {
    pub command: String,
    pub options: Vec<(String, Vec<String>)>,
    pub positionals: Vec<(String, String)>,
    pub leftovers: Vec<String>,
}

pub fn choose_command<'a>(spec: &'a Spec, argv: &[String]) -> &'a Command {
    if argv.is_empty() {
        return &spec.commands[0];
    }

    let first_arg = &argv[0];
    
    for cmd in &spec.commands {
        if let Some(lit) = first_literal(cmd) {
            if lit == *first_arg {
                return cmd;
            }
        }
    }

    &spec.commands[0]
}

fn first_literal(cmd: &Command) -> Option<String> {
    for item in &cmd.items {
        if let Item::Required(group) = item {
            match group {
                Group::Single(Atom::Lit(s)) => return Some(s.clone()),
                Group::Alt(atoms) => {
                    if let Some(Atom::Lit(s)) = atoms.first() {
                        return Some(s.clone());
                    }
                }
                _ => {}
            }
        }
    }
    None
}

pub fn parse_with(cmd: &Command, argv: &[String]) -> Result<ParseResult, ArgError> {
    let seq: Vec<(bool, &Group)> = cmd
        .items
        .iter()
        .map(|item| match item {
            Item::Required(g) => (false, g),
            Item::Optional(g) => (true, g),
        })
        .collect();

    let atoms: Vec<&Atom> = seq.iter().flat_map(|(_, g)| g.atoms()).collect();

    let known_opts: Vec<String> = atoms
        .iter()
        .flat_map(|atom| match atom {
            Atom::OptBool { long, short } => {
                let mut v = vec![];
                if let Some(l) = long {
                    v.push(l.clone());
                }
                if let Some(s) = short {
                    v.push(s.clone());
                }
                v
            }
            Atom::OptVal { long, short, .. } => {
                let mut v = vec![];
                if let Some(l) = long {
                    v.push(l.clone());
                }
                if let Some(s) = short {
                    v.push(s.clone());
                }
                v
            }
            _ => vec![],
        })
        .collect();

    let pos_list: Vec<(String, Type, Option<String>)> = atoms
        .iter()
        .filter_map(|atom| match atom {
            Atom::Pos { name, ty, default } => Some((name.clone(), ty.clone(), default.clone())),
            _ => None,
        })
        .collect();

    let mut opts_map: Vec<(String, Vec<String>)> = vec![];
    let mut pos_map: Vec<(String, String)> = vec![];
    let mut seen_pos = 0;
    let mut leftovers: Vec<String> = vec![];

    let args = argv.to_vec();
    let mut i = 0;

    while i < args.len() {
        let arg = &args[i];

        if arg == "--" {
            // Handle remaining positionals
            let remaining_pos = &pos_list[seen_pos..];
            let rest = &args[i + 1..];
            
            for (j, (name, ty, default)) in remaining_pos.iter().enumerate() {
                if j < rest.len() {
                    let val = parse_val(ty, &rest[j])?;
                    pos_map.push((name.clone(), val));
                    seen_pos += 1;
                } else if let Some(d) = default {
                    let val = parse_val(ty, d)?;
                    pos_map.push((name.clone(), val));
                    seen_pos += 1;
                } else {
                    return Err(ArgError(format!("Missing positional: {}", name)));
                }
            }
            
            leftovers = rest[remaining_pos.len()..].to_vec();
            break;
        } else if arg.starts_with('-') {
            let (name_part, val_part) = if let Some(pos) = arg.find('=') {
                let (n, v) = arg.split_at(pos);
                (n.to_string(), Some(&v[1..]))
            } else {
                (arg.clone(), None)
            };

            if !known_opts.contains(&name_part) {
                return Err(ArgError(format!("Unknown option: {}", name_part)));
            }

            let opt_decl = atoms
                .iter()
                .find(|atom| match atom {
                    Atom::OptVal { long, short, .. } => {
                        Some(&name_part) == long.as_ref() || Some(&name_part) == short.as_ref()
                    }
                    Atom::OptBool { long, short } => {
                        Some(&name_part) == long.as_ref() || Some(&name_part) == short.as_ref()
                    }
                    _ => false,
                })
                .ok_or_else(|| ArgError("Unknown option declaration".to_string()))?;

            match opt_decl {
                Atom::OptBool { .. } => {
                    add_opt(&mut opts_map, &name_part, "true".to_string());
                    i += 1;
                }
                Atom::OptVal { ty, .. } => {
                    let val = match val_part {
                        Some(v) => parse_val(ty, v)?,
                        None => {
                            if i + 1 < args.len() {
                                i += 1;
                                parse_val(ty, &args[i])?
                            } else {
                                return Err(ArgError(format!("Missing value for {}", name_part)));
                            }
                        }
                    };
                    add_opt(&mut opts_map, &name_part, val);
                    i += 1;
                }
                _ => return Err(ArgError("impossible".to_string())),
            }
        } else {
            // Try to consume as literal
            let mut consumed = false;
            for (req, group) in &seq {
                if !req && matches!(group, Group::Single(Atom::Lit(_)) | Group::Alt(_)) {
                    match group {
                        Group::Single(Atom::Lit(s)) if s == arg => {
                            consumed = true;
                            break;
                        }
                        Group::Alt(alts) => {
                            if alts.iter().any(|a| matches!(a, Atom::Lit(s) if s == arg)) {
                                consumed = true;
                                break;
                            }
                        }
                        _ => {}
                    }
                }
            }

            if consumed {
                i += 1;
                continue;
            }

            // Treat as positional
            if seen_pos >= pos_list.len() {
                return Err(ArgError(format!("Unexpected argument: {}", arg)));
            }

            let (name, ty, _) = &pos_list[seen_pos];
            let val = parse_val(ty, arg)?;
            pos_map.push((name.clone(), val));
            seen_pos += 1;
            i += 1;
        }
    }

    // Add option defaults
    for atom in &atoms {
        match atom {
            Atom::OptVal {
                long,
                short,
                ty,
                default,
                ..
            } => {
                let key = long.as_ref().or(short.as_ref()).unwrap();
                if !opts_map.iter().any(|(k, _)| k == key) {
                    if let Some(d) = default {
                        let val = parse_val(ty, d)?;
                        opts_map.push((key.clone(), vec![val]));
                    }
                }
            }
            Atom::OptBool { long, short } => {
                let key = long.as_ref().or(short.as_ref()).unwrap();
                if !opts_map.iter().any(|(k, _)| k == key) {
                    opts_map.push((key.clone(), vec!["false".to_string()]));
                }
            }
            _ => {}
        }
    }

    // Add positional defaults
    for (name, ty, default) in &pos_list {
        if !pos_map.iter().any(|(k, _)| k == name) {
            if let Some(d) = default {
                let val = parse_val(ty, d)?;
                pos_map.push((name.clone(), val));
            } else {
                return Err(ArgError(format!("Missing positional: {}", name)));
            }
        }
    }

    Ok(ParseResult {
        command: cmd.name.clone(),
        options: opts_map,
        positionals: pos_map,
        leftovers,
    })
}

fn parse_val(ty: &Type, s: &str) -> Result<String, ArgError> {
    match ty {
        Type::Int => {
            s.parse::<i64>()
                .map_err(|_| ArgError(format!("Expected INT, got: {}", s)))?;
            Ok(s.to_string())
        }
        Type::Bool => {
            let lower = s.to_lowercase();
            match lower.as_str() {
                "true" | "false" => Ok(lower),
                _ => Err(ArgError(format!("Expected BOOL (true|false), got: {}", s))),
            }
        }
        Type::Str | Type::Path => Ok(s.to_string()),
    }
}

fn add_opt(opts_map: &mut Vec<(String, Vec<String>)>, key: &str, val: String) {
    if let Some((_, vals)) = opts_map.iter_mut().find(|(k, _)| k == key) {
        vals.push(val);
    } else {
        opts_map.push((key.to_string(), vec![val]));
    }
}
